### ISSUE TITLE
Triaging issues and pulls

### ISSUE THREAD

**Issue by simison** [agenda/agenda#441]
It's a bit much for one person to go trough everything.

Any ideas/experiences how this should be done? Using tags or just list here (feel free to modify this post)?

Before anything is merged, should we have at least one "LGTM" from anyone in the team?

Everything should probably be sorted into:
- high priority
- low priority
- should be closed

**Comment by lushc** [comment]
Here are my opinions based on me poking around the repo the past few days and incorporating some of these PRs:

* #268 can be rebased and merged. There's a small conflict where the `lockOnTheFly()` query was refactored, [see this commit for reference](https://github.com/Codevate/agenda/commit/048c7f14e89e5fdcfe86dc74cc7aae42d7948bfe#diff-1ba7680cb21f30bb02af3c7899c41248L582). High priority.
* #241 needs more investigation. Although it [fixed the recursive lock issue for sshen81](https://github.com/agenda/agenda/issues/104#issuecomment-161750668) I still got errors when using the [benchmarking script](https://github.com/nwkeeley/agenda/blob/56013e472df6f08fa9ed0902aa2706f55dddd9bd/script.js) mentioned in #104, but maybe that's because the index order wasn't adjusted. That said, it also partially reverts the reasoning for the query changes in #180 and could be made redundant by the next PR. Low priority.
* #300 needs more work as I noticed [two problems after rebasing](https://github.com/agenda/agenda/pull/300#issuecomment-308211215) and it might benefit from a [query and index order tweak](https://github.com/Codevate/agenda/commit/d89b58c7cd43a78e39accb534eccb20c48fbb6b8). It completely solved #104 for me and potentially closes similar performance issues such as #437. High priority.
* #423 might be closed in favour of better documentation. The main issue is that using `repeatEvery()` in isolation doesn't set the `nextRunAt` time, but calling`computeNextRunAt()` doesn't work either since `lastRunAt` has yet to be set for new jobs, so the interval is ignored. [This code example](https://github.com/agenda/agenda/issues/192#issuecomment-140126659) could be added to the documentation as an example of delaying the execution of a repeat job. Low priority.
* #317 needs rebasing and is incomplete. It seems to have been left at the point where the contributor wasn't sure how to solve the issue of two different instances locking the same job, which does break the current assurance that multiple instances (forking, PM2 cluster etc.) can work from a single job queue without duplicating execution. It addresses #209 which, based on my limited testing, could instead by solved by documenting how to use the job API for this use-case (however, there may be some caveats to this approach that I'm unaware of): 

```js
const repeatJob = (name, interval, data) => {
  return new Promise((resolve, reject) => {
    agenda.create(name, data)
      .repeatEvery(interval)
      .save(resolve);
  });
};

const job1 = repeatJob('userJob', '1 second', {userId: 1});
const job2 = repeatJob('userJob', '5 seconds', {userId: 2});

Promise.all([job1, job2]);
```
* #433 can be rebased and merged. It seems to work fine for me, although I've not been relying on these values unlike Agendash. High priority?
* #408 I'm not sure of. It doesn't seem to have much of a use-case outside of the contributor's initial issue of using jobs as part of other models, and then not being able to determine if the job object was created by Agenda or not. Low priority.
* #340 can be closed as it was merged as part of #353.
* #418 is a useful reminder and should be merged.
* #383 I'm not sure of. I haven't come across this use-case myself yet but I'd imagine if you needed more context, couldn't `reason` just be an object you've passed in with all the attributes you need, since it can be stored as-is in MongoDB? Low priority.
* #361 I haven't really looked into, but it seems that the change of indentation is blocking things, I couldn't see any tests for the described scenarios either. Low priority.
* #321 is probably low priority but potentially worthwhile?
* #246 seems useful but needs a bit more work to incorporate some tests and fit the coding style to the rest of the library. Low priority.
* #166 is now out of date and should be refactored to use [this code example](https://github.com/agenda/agenda/pull/166#issuecomment-160111449) instead, there was a call for documentation and tests also. Low priority.

With the open PRs addressed it should be easier to then go through the more recent issues as you can just close whatever is duplicated/fixed by them. For older issues, I've seen other repos @ the reporter and asking them if this problem is still relevant, and if there's no further discussion after an arbitrary amount of time (30 days maybe?) then the issue is closed. Hope this helps.

**Comment by simison** [comment]
@lushc would you be interested in hopping on board and working on getting some of those high prio PRs merged in? Did you see #414?

Personally I have energy to poke around to get people active on this amazing library again but no really time (or insight in Agenda) to get coding.

**Comment by OmgImAlexis** [comment]
If someone can give me permission I'd like to go through and tag all issues so they can be handled easier.

Shall we close off any **QUESTION** issues that are over a year old? We can always reopen if the OP still has the issue.

**Comment by simison** [comment]
>Shall we close off any QUESTION issues that are over a year old? We can always reopen if the OP still has the issue.

+1

**Comment by simison** [comment]
@OmgImAlexis @lushc you should have rights to the repo now so feel free to start tackling stuff! :-) Feel free to assign PR's for yourself so it's clear who is working on what.

Me and @michelem09 have been closing/tagging old question issues.

**Comment by simison** [comment]
Anything related to "If server is shutdown, unfinished jobs won't resume" #410 seems like a very high prio to me. Any PR's that affect this?

**Comment by simison** [comment]
There's a really good question and possible bug at _Can't cancel a job once it's already locked_ #401

**Comment by OmgImAlexis** [comment]
I've added a few new labels for issues and closed off any that looked like questions but had no reply in over 12 months. Anything with a bug tag that has no-one assigned can be looked into.

**Comment by OmgImAlexis** [comment]
What's everyone think of creating a `develop` branch and switching the current PRs over to that so we can merge a little easier as we don't have to worry about everything working perfectly before releasing into master?

**Comment by simison** [comment]
@OmgImAlexis Good question. Personally I like more working with master as development branch (just all the tooling is built like that). 

That said I'm quite indifferent regarding branching so any people who are going to be more active in codebase in the future should decide. Unfortunately I can't promise to be super active code-wise here in the future but I'll do my best to push this so that it would have active team around it.

**Comment by OmgImAlexis** [comment]
We're down to [38 unlabeled issues](https://github.com/agenda/agenda/issues?page=1&q=is%3Aopen+is%3Aissue+no%3Alabel).

**Comment by OmgImAlexis** [comment]
I've added some milestones so all high priority PRs should be merged before 0.10.0 with all low priority expected be merged by 0.11.0 and the DB adapters waiting until 0.12.0

If you feel like they could be merged before or they are merged before please try and remember to change the milestone so the change log can be written from it.

#317 and #451 are still missing milestones. @lushc if you're free could you add whichever one you feel is appropriate for them.

**Comment by lushc** [comment]
@OmgImAlexis thanks, I'll take a look and see if I can add any more labels.

I'm not sure #317 can be assigned a milestone, I'll write a comment about this in the PR.

**Comment by OmgImAlexis** [comment]
So I've started on linting the whole project in https://github.com/agenda/agenda/pull/465 and I have [an issue](https://github.com/agenda/agenda/issues/455) open about possibly moving from eslint to xo, thoughts?

**Comment by emhagman** [comment]
@joeframbach I would also love access to begin working on this. We use this at our company and would love to be able to contribute.

**Comment by OmgImAlexis** [comment]
Adding onto @lushc's list I think we should work on adding promise support soon(next 1 - 3 months) and start deprecating the use of callbacks. Ref: https://github.com/agenda/agenda/issues/470

**Comment by simison** [comment]
@OmgImAlexis @lushc I moved #476 and #475 to milestone `1.0.0` which makes 0.10.1 almost ready now. You might want to move 0.11.0 after 1.0 or otherwise think about priorities again. :-)

**Comment by OmgImAlexis** [comment]
Sorry for being MiA for a day or so I'll be back tonight(5 - 10 hours). Had to deal with [bogus DMCA requests](https://github.com/github/dmca/blob/master/2017/2017-07-17-SiCKRAGE.md) on another project I work on. 😞

**Comment by niftylettuce** [comment]
The only core bug I currently see is <https://github.com/agenda/agenda/issues/401>.

**Comment by simison** [comment]
FYI, I renamed 0.11.0 to 1.1.0 and 0.12.0 to 1.2.0.

Now we have:
- 0.10.0 (almost ready)
- 1.0.0 (aka big breaking)
- 1.1.0
- 1.2.0

Remember never to be afraid to pump up the first digit if there's anything even remotely breaking going in. http://semver.org/

**Comment by simison** [comment]
I also added a few new "need" labels that I've found quite useful in other projects with many issues/PRs coming in:

![image](https://user-images.githubusercontent.com/87168/28513571-54a6c720-705f-11e7-8036-b59576fa4d10.png)

**Comment by OmgImAlexis** [comment]
Since I've merged the debug PR we now have 0.10.0 finished off. If there are no other changes that need to be made before a release then we should be good to go.

@simison can we get a release sorted? Still not sure who has npm permissions.

Also are we tagging our releases using git?

**Comment by OmgImAlexis** [comment]
If everyone is okay with this I'd like to get the xo PR added next before any of the breaking 1.0.0 changes go in so we know all future PRs will be 100% linted since the longer we leave it the more code we've got to switch over.

**Comment by OmgImAlexis** [comment]
@agenda/maintainers could some of the people added please check PRs. Seems over the last week I've only seen @simison @lushc @emhagman and @niftylettuce commenting and approving PRs and I'd like to get more PRs merged and the code cleaned up to a better state but I can't do anything while I'm sitting here waiting for PRs to be approved and I'd rather not just merge code without anyone else checking off on it. I get that some of you may be busy but there's 10 people in the maintainers group and 6 of them I haven't seen since they were added. 😞

**Comment by niftylettuce** [comment]
@OmgImAlexis I will be updating my PR #464 soon to remove eslint stuff and also fix all the other issues mentioned

**Comment by niftylettuce** [comment]
@OmgImAlexis I will have more time come next week - we are doing a major release from last 3 months of work so I apologize I cannot contribute more at this time - I have been actively reading all notifications though so I am in tune.

**Comment by OmgImAlexis** [comment]
@niftylettuce that's all good. Once we get 0.10.0 published I'd like to get your PR merged so we can start on the 1.0.0 changes.

**Comment by simison** [comment]
Sorry I forgot to share this info here; I emailed @rschmukler about npm access and waiting for a reply.

**Comment by OmgImAlexis** [comment]
@rschmukler can you please check your emails we can't really move forward with agenda until we get npm sorted. 😞

**Comment by simison** [comment]
@OmgImAlexis, @rschmukler actually did reply some days ago that he'll add you guys, but he also asked if he could add me and I've been busy to reply for some days so now it's been lagging due me. :-) I replied him yesterday so hopefully we'll get this sorted soon!

Patience. :-) Hope this doesn't affect your enthusiasm!

**Comment by OmgImAlexis** [comment]
I've published `0.10.0` and pushed the git tag since all issues in the `0.10.0` milestone were closed.

I suggest whoever handles the npm publishing use [np](https://www.npmjs.com/package/np) as this will keep our way of publishing consistent no matter who's doing it.

**Comment by OmgImAlexis** [comment]
Thoughts on dropping codeclimate since we now have xo and will soon have ava?

Since xo can already tell us the same stuff as codeclimate it's redundant now. Maybe if codeclimate ever adds in native xo support we can then switch to to using it.

**Comment by simison** [comment]
@OmgImAlexis could you add using `np` to wiki? https://github.com/agenda/agenda/wiki#generating-a-new-release Thanks!

**Comment by lushc** [comment]
I don't have a strong opinion either, but I'm all in favour of removing codeclimate if it's redundant 👍

**Comment by OmgImAlexis** [comment]
@simison [added](https://github.com/agenda/agenda/wiki#generating-a-new-release).

I'll remove codeclimate then. 👍

**Comment by simison** [comment]
@joeframbach see above

**Comment by OmgImAlexis** [comment]
We should look at updating agendash to be inline with the versions numbers here so we push a new `1.0.0` build of agendash when we push this to npm.

**Comment by OmgImAlexis** [comment]
After my stuff up of pushing master as `0.10.1` on npm I think we need to discuss using a `develop` branch. With some of the other projects I work on we use a master and develop branch with master being the live copy on npm, etc. and develop being the branch all changes are merge into.

This allows us to keep merging code without needing to worry if everything is 100% perfect as we do a final lot of tests before it's merged into master. This also makes hotfixing master really easy as we just merge the hotfix into master and bump the version without all the new changes being there.

At the moment the best way we can hot fix it to checkout a new branch that's a few commits behind and apply the hotfix to that and push to npm as I had to with `0.10.2`.

**Comment by OmgImAlexis** [comment]
@agenda/maintainers I know you're all probably busy but I'd really like to get agenda moving along. If you're all busy and are okay with handing merging without a `LGTM` from another member then I'll start doing that otherwise could you guys try to atleast check notifications from here a little more often..

**Comment by simison** [comment]
Yeah totally, progress is more important.

**Comment by OmgImAlexis** [comment]
Are all methods meant to be chainable? I noticed `agenda.cancel` doesn't return `this`.

https://github.com/agenda/agenda/blob/master/lib/agenda/cancel.js#L11-L23

**Comment by fzondlo** [comment]
We have started using agenda at my company as well. Please let me know if there's anything I can do to help. Would love to get involved.

**Comment by dandv** [comment]
Any progress on enabling more maintainers to join? 58 open PRs is quite a bit.
